spec/fastsync/version2/fsv2_scheduler.tla

```diff
@@ -0,0 +1,532 @@
+ 
+ The model includes:
+    - a scheduler that maintains the peers and blocks that it receives from the peers, and
+    - one environment simulating a correct peer
+    
+ This specification focuses on the events that are received and produced by the scheduler.
+ Communication between the scheduler and the other fastsync components is not specified.
+*)
+
+EXTENDS Integers, FiniteSets, TLC
+
+\* the protocol parameters
+CONSTANTS
+    PeerIDs,        \* potential peer ids, a set of integers, e.g., 0..2
+    ultimateHeight, \* the maximum height of the blockchain, an integer, e.g., 3
+    numRequests     \* the number of requests made per batch, e.g., 2
+
+\* a few definitions
+None == -1                      \* an undefined value
+Heights == 1..ultimateHeight    \* potential heights
+noErr == "errNone"
+Errors == {
+  noErr, "errPeerNotFound", "errDelRemovedPeer", "errAddDuplicatePeer",
+  "errUpdateRemovedPeer", "errAfterPeerRemove", "errBadPeer", "errBadPeerState",
+  "errProcessedBlockEv", "finished"}
+
+PeerStates == {"peerStateUnknown", "peerStateNew", "peerStateReady", "peerStateRemoved"}
+
+BlockStates == {
+  "blockStateUnknown", "blockStateNew", "blockStatePending",
+  "blockStateReceived", "blockStateProcessed"}
+
+\* basic stuff
+Min(a, b) == IF a < b THEN a ELSE b
+Max(a, b) == IF a > b THEN a ELSE b
+
+\* the state of the scheduler:
+VARIABLE  turn          \* who makes a step: the scheduler or the environment
+
+\* the state of the reactor: 
+VARIABLES inEvent,       \* an event from the environment to the scheduler
+          envRunning \* a Boolean, negation of stopProcessing in the implementation
+                    
+\* the state of the scheduler:
+VARIABLES outEvent,       \* an event from the scheduler to the environment
+          scRunning
+
+\* the block pool: 
+VARIABLE scheduler
+    (*
+       scheduler is a record that contains: 
+         height: Int,
+            height of the next block to collect
+         peers: PeerIDs,
+            the set of active peers, not known to be faulty
+         peerHeights: [PeerIDs -> Heights],
+            a map to collect the peer responses about their heights 
+         peerStates: [PeerIDs -> PeerStates],
+            a map to record the peer states 
+         blockStates: [Heights -> BlockStates]
+            a set of heights for which blocks are to be scheduled, pending or received
+         pendingBlocks: [Heights -> PeerIDs],
+            a set of heights for which blocks are to be scheduled or pending
+         receivedBlocks: [Heights -> PeerIDs],
+            a set of heights for which blocks were received but not yet processed
+    *)
+    
+vars == <<turn, envRunning, inEvent, scRunning, outEvent, scheduler>>    
+
+\* for now just keep the height in the block
+Blocks == [ height: Heights ]
+
+noEvent == [type |-> "NoEvent"] 
+
+InEvents == 
+  {noEvent} \cup
+  [type: {"rTrySchedule", "tNoAdvanceExp", "rTryPrunePeer"}] \cup
```

Comment:
Looks like we're not actually using the rTryPrunePeer anywhere?

Reply:
No, not yet implemented...it's just a bunch of RemovePeer-s for unresponsive peers so almost already modeled but will add it soon.